error_budget.go

```diff
@@ -0,0 +1,39 @@
+package hostpool
+
+import (
+	"time"
+)
+
+type ringBuffer struct {
+	size  int
+	index int
+	items []time.Time
+}
+
+func NewRingBuffer(size int) *ringBuffer {
+	return &ringBuffer{
+		size:  size,
+		items: make([]time.Time, size),
+	}
+}
+
+func (r *ringBuffer) insert(ts time.Time) {
+	r.index = (r.index + 1) % r.size
```

Comment:
I think this works in usages of `since` because we walk all entries, but just reading this again is the scenario that on the _first_ insert you'd have:

`r.index` is initialised to 0, we then increment by 1 and mod size, then set an element, meaning `r.items` will be `[time.Time{}, ts, time.Time{},…]` (i.e., that the first item inserts to item 1 rather than item 0)? It seems a little surprising—perhaps the order should be to write at the cursor, then shift the cursor.

Reply:
Yeah, originally the code in `FailuresSince` used the index, so it was convenient for it to be set to the last inserted item. This is vestigial from that. I agree it's not ideal but I think we can leave it for the moment and improve in future.